pkg/controller/bundle_sync_task.go

```diff
@@ -102,6 +109,80 @@ func (st *bundleSyncTask) process() (retriableError bool, e error) {
 	return false, nil
 }
 
+// Process the bundle marked with DeletionTimestamp
+// TODO: remove this method after https://github.com/kubernetes/kubernetes/issues/59850 is fixed
+func (st *bundleSyncTask) processDeleted() (retriableError bool, e error) {
+	if hasDeleteResourcesFinalizer(st.bundle) {
+		if !HasFinalizer(st.bundle, meta_v1.FinalizerDeleteDependents) {
+			// If "foregroundDeletion" finalizer was not set, perform manual cascade deletion
+			retrieable, err := st.deleteAllResources()
+			if err != nil {
+				return retrieable, err
+			}
+		}
+
+		// If the "foregroundDeletion" finalizer is set, or the manual deletion
+		// of resources has succeeded, remove the "deleteResources" finalizer
+		st.newFinalizers = removeDeleteResourcesFinalizer(st.bundle.GetFinalizers())
+	}
+	return false, nil
+}
+
+// TODO: delete this method after the issue in kubectl is fixed:
+func (st *bundleSyncTask) deleteAllResources() (retriableError bool, e error) {
+	objs, err := st.store.ObjectsControlledBy(st.bundle.Namespace, st.bundle.UID)
+	if err != nil {
+		return false, err
+	}
+	existingObjs := make(map[objectRef]types.UID, len(objs))
+	for _, obj := range objs {
+		m := obj.(meta_v1.Object)
+		if m.GetDeletionTimestamp() != nil {
+			// Object is marked for deletion already
+			continue
+		}
+		ref := objectRef{
+			GroupVersionKind: obj.GetObjectKind().GroupVersionKind(),
+			Name:             m.GetName(),
+		}
+		existingObjs[ref] = m.GetUID()
+	}
+	var firstErr error
+	retriable := true
+	policy := meta_v1.DeletePropagationForeground
+	for ref, uid := range existingObjs {
```

Comment:
What is the reason for two loops?

Reply:
This method is a subset of `deleteRemovedResources`. Can simplify a bit ofc.